---
title: "Seurat_object_creation"
author: "MOHANKumar"
date: "2024-01-02"
output: html_document
---

```{r setup, include=FALSE,eval=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```




# Libraries 

```{r,warning = FALSE, message = FALSE,eval=FALSE}
library(Seurat)
library(dplyr)
library(tidyverse)
library(SeuratWrappers)
library(ggplot2)
```


# Standard Pre-Processing Work Flow:
The steps below encompass the standard pre-processing workflow for scRNA-seq data in Seurat. The seurat object is created for 5 different file types. This workflows focus on creating the seurat object for further analyis.



```{r,warning = FALSE, message = FALSE,eval=FALSE}

# reading the files 
IPF_I_mtx <- ReadMtx(mtx = "D:/A_sc_trajectory/IPF1L/GSM3891626_IPF1_Lin_matrix.mtx.gz",
                 features = "D:/A_sc_trajectory/IPF1L/GSM3891626_IPF1_Lin_features.tsv.gz",
                 cells = "D:/A_sc_trajectory/IPF1L/GSM3891626_IPF1_Lin_barcodes.tsv.gz")


IPF_II_mtx <- ReadMtx(mtx = "D:\\A_sc_trajectory\\IPF2L\\GSM3891628_IPF2_Lin_matrix.mtx.gz",
                   features = "D:\\A_sc_trajectory\\IPF2L\\GSM3891628_IPF2_Lin_featurestsv.gz",
                   cells = "D:\\A_sc_trajectory\\IPF2L\\GSM3891628_IPF2_Lin_barcodes.tsv.gz")

IPF_III_mtx <- ReadMtx(mtx = "D:\\A_sc_trajectory\\IPF3L\\GSM3891630_IPF3_Lin_matrix.mtx.gz",
                   features = "D:\\A_sc_trajectory\\IPF3L\\GSM3891630_IPF3_Lin_features.tsv.gz",
                   cells = "D:\\A_sc_trajectory\\IPF3L\\GSM3891630_IPF3_Lin_barcodes.tsv.gz")

# Create a Seurat  object
IPF_I <- CreateSeuratObject(counts = IPF_I_mtx, project = "IPF1", min.cells = 3, min.features = 200)

IPF_II <- CreateSeuratObject(counts = IPF_II_mtx, project = "IPF2", min.cells = 3, min.features = 200)

IPF_III <- CreateSeuratObject(counts = IPF_III_mtx, project = "IPF3", min.cells = 3, min.features = 200)

```


## **Step 1: QC and selecting cells for further analysis**

  The number of unique genes detected in each cell. Similarly, the total number of molecules detected within a cell (correlates stronglywith unique genes).The percentage of reads that map to the mitochondrial genome.If the percentage of mitochondrial gene apply filter to remove the hight percentage genes.

```{r,warning = FALSE, message = FALSE,eval=FALSE}
IPF_I <- PercentageFeatureSet(IPF_I,pattern = "^MT-",col.name = "percent.mnt")
IPF_II <- PercentageFeatureSet(IPF_II,pattern = "^MT-",col.name = "percent.mnt")
IPF_III <- PercentageFeatureSet(IPF_III,pattern = "^MT-",col.name = "percent.mnt")

IPF <- merge(IPF_I,y=c(IPF_II,IPF_III))
```

# Plots for the gene expression 

```{r,warning = FALSE, message = FALSE,eval=FALSE}
v1 <- VlnPlot(IPF,features = c("nFeature_RNA", "nCount_RNA", "percent.mnt"),ncol = 3)
print(v1)


 scatter_plot <- ggplot(IPF@meta.data, aes(x = nCount_RNA, y = nFeature_RNA, color =orig.ident)) +
   geom_point() +
   ggtitle("Scatter Plot of nCount_RNA vs nFeature_RNA") +
   geom_smooth(method = "lm", se = FALSE,color = "black") +
   labs(x = "nCount_RNA", y = "nFeature_RNA")
 
 print(scatter_plot)

```


# Filter the Low quality genes

```{r,warning = FALSE, message = FALSE,eval=FALSE}
IPF <- subset(IPF, subset = nFeature_RNA < 4500 & nCount_RNA < 20000 & 
                            percent.mnt < 10)
```


# Plot for gene expression

```{r,warning = FALSE, message = FALSE,eval=FALSE}
v1 <- VlnPlot(IPF,features = c("nFeature_RNA", "nCount_RNA", "percent.mnt"),ncol = 3)
print(v1)


 scatter_plot <- ggplot(IPF@meta.data, aes(x = nCount_RNA, y = nFeature_RNA, color =orig.ident)) +
   geom_point() +
   ggtitle("Scatter Plot of nCount_RNA vs nFeature_RNA") +
   geom_smooth(method = "lm", se = FALSE,color = "black") +
   labs(x = "nCount_RNA", y = "nFeature_RNA")
 
 print(scatter_plot)

```


# **Step 2: Normalizing the data**
  
  After removing unwanted cells from the dataset, the next step is to normalize the data. By default, we employ a global-scaling normalization method “LogNormalize” that normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result.
  
```{r,warning = FALSE, message = FALSE,eval=FALSE}
IPF.list <- SplitObject(IPF,split.by ="orig.ident" )

IPF.list <-lapply(IPF.list,FUN = function(x){
  x<- NormalizeData(x)
  x <- FindVariableFeatures(x,selection.method="vst",nfeatures = 2000)
})
```


### Finding the Anchors

```{r,warning = FALSE, message = FALSE,eval=FALSE}

features <- SelectIntegrationFeatures(object.list = IPF.list)
IPF.anchors <- FindIntegrationAnchors(object.list = IPF.list,anchor.features = features)

# creates an integrated data assay 
IPF.combined <- IntegrateData(anchorset = IPF.anchors)
DefaultAssay(IPF.combined)<- "integrated"

integrated_data <-IPF.combined@assays$integrated
anchor_genes <- rownames(integrated_data)
```

# **Step 3:Identification of highly variable features (Feature Selection)**

  We next calculate a subset of features that exhibit high cell-to-cell variation in the dataset (i.e, they are highly expressed in some cells, and lowly expressed in others). We and others have found that focusing on these genes in downstream analysis helps to highlight biological signal in single-cell datasets.
  
  


# **Step 4: Scaling the data**

  Next, we apply a linear transformation (‘scaling’) that is a standard pre-processing step prior to dimensional reduction techniques like PCA


```{r,warning = FALSE, message = FALSE,eval=FALSE}
IPF.combined <- ScaleData(IPF.combined,verbose = FALSE)
```


# **Step 5: Perform Linear dimensional reduction**

  Next we perform PCA on the scaled data. By default, only the previously determined variable features are used as input, but can be defined using features argument if you wish to choose a different subset (if you do want to use a custom subset of features, make sure you pass these to ScaleData first).
  
```{r,warning = FALSE, message = FALSE,eval=FALSE}
IPF.combined <- RunPCA(IPF.combined,npcs = 30,verbose = FALSE)
```

# **Step 6: Determine the ‘dimensionality’ of the dataset**

  To overcome the extensive technical noise in any single feature for scRNA-seq data, Seurat clusters cells based on their PCA scores, with each PC essentially representing a ‘metafeature’ that combines information across a correlated feature set. The top principal components therefore represent a robust compression of the dataset.
  
```{r,warning = FALSE, message = FALSE,eval=FALSE}

ElbowPlot(IPF.combined)


```

**Step 7: Cluster the cell**

Partitioning the cellular distance matrix into clusters 

```{r,warning = FALSE, message = FALSE,eval=FALSE}
IPF.combined <- FindNeighbors(IPF.combined,reduction = "pca",dims = 1:30)
IPF.combined <- FindClusters(IPF.combined,resolution = 0.5)
```


# **Naming the clusters**

```{r,warning = FALSE, message = FALSE,eval=FALSE}

IPF.subset_coimbine <- RenameIdents(IPF.combined,
                                     `0`="Alveolar",
                                     `1`="Alveolar",
                                     `2` = "Alveolar",
                                     `3`='Peribronchia',
                                     `4`='Adventitia',
                                     `5`='SMC',
                                     `6`='Adventitial',
                                     `7`='SMC',
                                     `8`='Adventitia',
                                     `9`='Pericytes',
                                     `10`= "Alveolar")
```



#**step 8: Run non-linear dimensional reduction (UMAP/tSNE)**

```{r,warning = FALSE, message = FALSE,eval=FALSE}
IPF.combined <- RunUMAP(IPF.combined,reduction = "pca",dims = 1:30)
```



```{r,warning = FALSE, message = FALSE,eval=FALSE}
DimPlot(IPF.combined,reduction="umap",label= TRUE)
```




**saving the Processed File**
```{r,warning = FALSE, message = FALSE,eval=FALSE}
saveRDS(IPF.combined, file = r"(D:\A\App\processed_data.rds)")
```





